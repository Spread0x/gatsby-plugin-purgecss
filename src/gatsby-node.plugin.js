const path = require('path');
const PurgecssPlugin = require('purgecss-webpack-plugin');
const glob = require('glob')

const PATHS = {
  src: path.join(process.cwd(), 'src/**/!(*.d).{ts,js,jsx,tsx}')
};

exports.onCreateWebpackConfig = (
  { actions, stage, rules, plugins, loaders },
  { postCssPlugins, ...userOptions }
) => {
  // if (stage !== 'build-javascript') {
  //   return;
  // }

  const { setWebpackConfig } = actions;

  userOptions = {
    // paths: [PATHS.src],
    whitelist: ['html', 'body'],
    ...userOptions
  };

  console.log(plugins);

  const rule = {
    // test: /stylus\.(s?(a|c)ss|less|styl)$/,
    test: /\.styl$/,
    use: [
      loaders.miniCssExtract(),
      loaders.style(),
      loaders.css({ importLoaders: 2 }),
      loaders.postcss({ plugins: postCssPlugins }),
      {
        loader: path.join(__dirname, 'loader.js'),
        options: userOptions
      }
    ]
  };

  setWebpackConfig({
    plugins: [
      new PurgecssPlugin({
        paths: glob.sync(PATHS.src, { nodir: true }),
        ...userOptions
      })
    ]
  });
};
