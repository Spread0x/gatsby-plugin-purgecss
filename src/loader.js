// modified from https://github.com/americanexpress/purgecss-loader

const PurgeCss = require('purgecss')
const { getOptions } = require('loader-utils')

module.exports = function loader(source) {
  console.log("\nPURGE\n")
  const options = getOptions(this)

  const css = new PurgeCss({
    css: [{ raw: source }],
    ...options,
  }).purge()

  if (options.rejected) {
    const rejected = css[0].rejected

    if (rejected && rejected.length) {
      const filtered = rejected.map((val) => {
        return val.replace('\n', '')
      })
      console.log('\nRemoved Selectors: ', filtered)

    }
  }

  return css[0].css
}
