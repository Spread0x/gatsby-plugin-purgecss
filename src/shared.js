class Logger {
  constructor() {
    this.removedQuantity = 0;
  }

  add(quantity) {
    this.removedQuantity = this.removedQuantity + quantity;
  }
}

const logger = new Logger();

export default logger;
