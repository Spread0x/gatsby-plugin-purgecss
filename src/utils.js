const findLoader = (useArray, loaderNamesRegex) => {
  const index = useArray.findIndex(loaderObject => {
    let loaderName = '';

    if (typeof loaderObject === 'string') {
      loaderName = loaderObject;
    } else {
      loaderName = loaderObject.loader || '';
    }

    return loaderName.match(loaderNamesRegex);
  });

  return index === -1 ? null : index;
};

const insertLoader = (useArray, index, loader) => {
  if (index === null) {
    return
  }
  useArray.splice(index, 0, loader);
};

export { findLoader, insertLoader };
