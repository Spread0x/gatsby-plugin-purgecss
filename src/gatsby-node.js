/**
 * @typedef {import('webpack').Configuration} WebpackConfig
 * @typedef {import('webpack').RuleSetRule} RuleSetRule
 * @typedef {import('webpack').RuleSetLoader} RuleSetLoader
 */

import { findLoader, insertLoader, color } from './utils';
import path from './paths';
import { stats } from './shared';

const loadersRegex = /stylus-loader|sass-loader|less-loader/;

export function onCreateWebpackConfig(
  { actions, stage, getConfig },
  { plugins, ...userOptions }
) {
  userOptions = {
    content: [path.src],
    rejected: true,
    printRejected: false,
    whitelist: ['html', 'body'],
    ...userOptions
  };

  if (userOptions.rejected && stage === 'build-html') {
    console.log(color.FgGreen, `\ngatsby-plugin-purgecss:`);
    console.log(color.FgCyan, `Previous CSS Size: ${stats.oldSize} KB`);
    console.log(
      color.FgCyan,
      `New CSS Size: ${stats.newSize} KB (-${stats.percentChange}%)`
    );
    console.log(color.FgYellow, `Removed ~${stats.sizeDifference} KB of CSS`);
    console.log(color.Reset);
  }

  if (stage !== 'build-javascript') {
    return;
  }

  const { setWebpackConfig } = actions;

  /**
   * @type {WebpackConfig}
   */
  const config = getConfig();
  const existingRules = config.module.rules;

  /**
   * @type {RuleSetLoader}
   */
  const purgecssloader = {
    loader: path.loader,
    options: userOptions
  };

  existingRules.forEach(rule => {
    if (Array.isArray(rule.oneOf)) {
      rule.oneOf.forEach(rule => {
        if (Array.isArray(rule.use)) {
          const index = findLoader(rule.use, loadersRegex);
          insertLoader(rule.use, index, purgecssloader);
        }
      });
    }
  });

  actions.replaceWebpackConfig(config);

  /**
   * @type {RuleSetRule}
   */
  const purgecssRule = {
    test: /\.css$/,
    use: [purgecssloader]
  };

  setWebpackConfig({
    module: {
      rules: [purgecssRule]
    }
  });
}
