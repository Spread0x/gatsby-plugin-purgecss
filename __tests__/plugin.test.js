import { onCreateWebpackConfig } from '../src/gatsby-node'
import configData from '../testData'

describe(`gatsby-plugin-pugrecss`, () => {
  const actions = {
    setWebpackConfig: jest.fn(),
    replaceWebpackConfig: jest.fn()
  };

  const getConfig = jest.fn().mockReturnValue(configData)

  beforeEach(() => {
    actions.setWebpackConfig.mockReset();
  });

  let stage = ['build-javascript', 'build-html'];
  const options = {
    plugins: []
  }

  it(`build-javascript: Should match snapshot`, () => {
    onCreateWebpackConfig({ actions, stage:stage[0], getConfig }, options);
    expect(actions.replaceWebpackConfig).toMatchSnapshot();
    expect(actions.setWebpackConfig).toMatchSnapshot();
  });

  it(`build-html: Should match snapshot`, () => {
    onCreateWebpackConfig({ actions, stage:stage[1], getConfig }, options);
    expect(actions.replaceWebpackConfig).toMatchSnapshot();
    expect(actions.setWebpackConfig).toMatchSnapshot();
  });
});
