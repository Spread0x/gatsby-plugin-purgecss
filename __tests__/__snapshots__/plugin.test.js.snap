// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gatsby-plugin-pugrecss build-html: Should match snapshot 1`] = `[MockFunction]`;

exports[`gatsby-plugin-pugrecss build-html: Should match snapshot 2`] = `[MockFunction]`;

exports[`gatsby-plugin-pugrecss build-javascript: Should match snapshot 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "exclude": /\\(node_modules\\|bower_components\\)/,
              "test": /\\\\\\.jsx\\?\\$/,
              "use": Array [
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/babel-loader/lib/index.js",
                  "options": Object {
                    "babelrc": false,
                    "cacheDirectory": true,
                    "plugins": Array [
                      "gatsby-plugin-purgecss/node_modules/babel-plugin-remove-graphql-queries/index.js",
                      Array [],
                      Array [],
                      Array [],
                    ],
                    "presets": Array [
                      Array [],
                      Array [],
                      Array [],
                    ],
                    "sourceType": "unambiguous",
                  },
                },
              ],
            },
            Object {
              "test": /\\\\\\.ya\\?ml/,
              "use": Array [
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/json-loader/index.js",
                  "options": Object {},
                },
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/yaml-loader/index.js",
                  "options": Object {},
                },
              ],
            },
            Object {
              "test": /\\\\\\.\\(eot\\|otf\\|ttf\\|woff\\(2\\)\\?\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
              "use": Array [
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/url-loader/dist/cjs.js",
                  "options": Object {
                    "limit": 10000,
                    "name": "static/[name]-[hash].[ext]",
                  },
                },
              ],
            },
            Object {
              "test": /\\\\\\.\\(ico\\|svg\\|jpg\\|jpeg\\|png\\|gif\\|webp\\)\\(\\\\\\?\\.\\*\\)\\?\\$/,
              "use": Array [
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/url-loader/dist/cjs.js",
                  "options": Object {
                    "limit": 10000,
                    "name": "static/[name]-[hash].[ext]",
                  },
                },
              ],
            },
            Object {
              "test": /\\\\\\.\\(mp4\\|webm\\|wav\\|mp3\\|m4a\\|aac\\|oga\\|flac\\)\\$/,
              "use": Array [
                Object {
                  "loader": "gatsby-plugin-purgecss/node_modules/url-loader/dist/cjs.js",
                  "options": Object {
                    "name": "static/[name]-[hash].[ext]",
                  },
                },
              ],
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.css\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.css\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
              ],
            },
            Object {
              "test": /HashHistory/,
              "use": "null-loader",
            },
            Object {
              "test": /MemoryHistory/,
              "use": "null-loader",
            },
            Object {
              "test": /StaticRouter/,
              "use": "null-loader",
            },
            Object {
              "test": /MemoryRouter/,
              "use": "null-loader",
            },
            Object {
              "test": /HashRouter/,
              "use": "null-loader",
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.styl\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/stylus-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
                Object {
                  "exclude": /\\\\\\.module\\\\\\.styl\\$/,
                  "test": /\\\\\\.styl\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/stylus-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
              ],
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/sass-loader/lib/loader.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.s\\(a\\|c\\)ss\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/sass-loader/lib/loader.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
              ],
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.less\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/less-loader/dist/cjs.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
                Object {
                  "test": /\\\\\\.less\\$/,
                  "use": Array [
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/mini-css-extract-plugin/dist/loader.js",
                      "options": Object {},
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/css-loader/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/postcss-loader/lib/index.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                    Object {
                      "loader": "path/loader.js",
                      "options": Object {
                        "content": Array [
                          "src/**/!(*.d).{ts,js,jsx,tsx}",
                        ],
                        "whitelist": Array [
                          "html",
                          "body",
                        ],
                      },
                    },
                    Object {
                      "loader": "gatsby-plugin-purgecss/node_modules/less-loader/dist/cjs.js",
                      "options": Array [
                        [Function],
                      ],
                    },
                  ],
                },
              ],
            },
            Object {
              "oneOf": Array [
                Object {
                  "test": /\\\\\\.module\\\\\\.test\\$/,
                  "use": "null-loader",
                },
                Object {
                  "test": /\\\\\\.test\\$/,
                  "use": "null-loader",
                },
              ],
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;

exports[`gatsby-plugin-pugrecss build-javascript: Should match snapshot 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "module": Object {
          "rules": Array [
            Object {
              "test": /\\.css\\$/,
              "use": Array [
                Object {
                  "loader": "path/loader.js",
                  "options": Object {
                    "content": Array [
                      "src/**/!(*.d).{ts,js,jsx,tsx}",
                    ],
                    "whitelist": Array [
                      "html",
                      "body",
                    ],
                  },
                },
              ],
            },
          ],
        },
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
